pontoon/base/views.py

```diff
@@ -427,6 +433,106 @@ def entities(request):
     }, safe=False)
 
 
+@login_required(redirect_field_name='', login_url='/403')
+@require_POST
+@require_AJAX
+@transaction.atomic
+def batch_edit_translations(request):
+    try:
+        l = request.POST['locale']
+        action = request.POST['action']
+        entity_pks = request.POST.getlist('entities[]')
+    except MultiValueDictKeyError as e:
+        return HttpResponseBadRequest('Bad Request: {error}'.format(error=e))
+
+    locale = get_object_or_404(Locale, code=l)
+
+    # Batch editing is only available to translators
+    if not request.user.has_perm('base.can_translate_locale', locale):
+        return HttpResponseForbidden(
+            "Forbidden: You don't have permission for batch editing"
+        )
+
+    entities = (
+        Entity.objects.filter(pk__in=entity_pks)
+        .prefetch_resources_translations(locale)
+    )
+
+    translation_pks = set()
+
+    for entity in entities:
+        if entity.string_plural == "":
+            translation_pks.add(entity.get_translation()['pk'])
+
+        else:
+            for plural_form in range(0, locale.nplurals or 1):
+                translation_pks.add(entity.get_translation(plural_form)['pk'])
+
+    translation_pks.discard(None)
+    translations = Translation.objects.filter(pk__in=translation_pks)
+
+    if action == 'approve':
+        translations = translations.filter(approved=False)
+        count = translations.count()  # Must come before update
+        translatedresources = translations.translated_resources()
+
+        translations.update(
+            approved=True,
+            approved_user=request.user,
+            approved_date=timezone.now()
+        )
+
+    elif action == 'delete':
+        count = translations.count()  # Must come before delete
+        translatedresources = translations.translated_resources()
+        translations.delete()
+
+    elif action == 'replace':
+        find = request.POST.get('find')
+        replace = request.POST.get('replace')
+
+        # Empty translations produced by replace might not be always allowed
+        forbidden = (
+            translations.filter(string=find)
+            .exclude(entity__resource__format__in=['properties', 'ini', 'dtd'])
+        )
+
+        if not replace and forbidden.exists():
+            return JsonResponse({
+                'error': 'Empty translations not allowed',
+            })
+
+        # Find translations
+        translations = translations.filter(string__contains=find)
```

Comment:
I guess we could use defer() here, because initialization of instances of models cost a lot of cpu/memory. Alternatively  we could use values() and pack these values into Translation() objects. We could test which way will be faster.